UefiTestingPkg/FunctionalSystemTests/MemoryProtectionTest/App/SmmMemoryProtectionTestApp.c

```diff
@@ -0,0 +1,790 @@
+  }
+
+  // Signal MM
+  if (!EFI_ERROR (Status)) {
+    Status = SmmCommunication->Communicate (
+                                 SmmCommunication,
+                                 CommHeader,
+                                 &CommBufferSize
+                                 );
+    DEBUG ((DEBUG_INFO, "%a - Communicate() = %r\n", __FUNCTION__, Status));
+  }
+
+  return VerificationCommBuffer->Status;
+}
+
+/**
+  Locates and stores address of comm buffer
+**/
+STATIC
+VOID
+LocateSmmCommonCommBuffer (
+  VOID
+  )
+{
+  EFI_STATUS                               Status;
+  EDKII_PI_SMM_COMMUNICATION_REGION_TABLE  *PiSmmCommunicationRegionTable;
+  EFI_MEMORY_DESCRIPTOR                    *SmmCommMemRegion;
+  UINTN                                    Index, BufferSize;
+
+  if (mPiSmmCommonCommBufferAddress == NULL) {
+    Status = EfiGetSystemConfigurationTable (&gEdkiiPiSmmCommunicationRegionTableGuid, (VOID **)&PiSmmCommunicationRegionTable);
+    if (EFI_ERROR (Status)) {
+      return;
+    }
+
+    // Only need a region large enough to hold a MEMORY_PROTECTION_TEST_COMM_BUFFER
+    BufferSize       = 0;
+    SmmCommMemRegion = (EFI_MEMORY_DESCRIPTOR *)(PiSmmCommunicationRegionTable + 1);
+    for (Index = 0; Index < PiSmmCommunicationRegionTable->NumberOfEntries; Index++) {
+      if (SmmCommMemRegion->Type == EfiConventionalMemory) {
+        BufferSize = EFI_PAGES_TO_SIZE ((UINTN)SmmCommMemRegion->NumberOfPages);
+        if (BufferSize >= (sizeof (MEMORY_PROTECTION_TEST_COMM_BUFFER) + OFFSET_OF (EFI_SMM_COMMUNICATE_HEADER, Data))) {
+          break;
+        }
+      }
+
+      SmmCommMemRegion = (EFI_MEMORY_DESCRIPTOR *)((UINT8 *)SmmCommMemRegion + PiSmmCommunicationRegionTable->DescriptorSize);
+    }
+
+    if (SmmCommMemRegion->PhysicalStart > MAX_UINTN) {
+      return;
+    }
+
+    mPiSmmCommonCommBufferAddress = (VOID *)(UINTN)SmmCommMemRegion->PhysicalStart;
+    mPiSmmCommonCommBufferSize    = BufferSize;
+  }
+}
+
+/**
+  This dummy function definition is used to test no-execute protection on allocated buffers
+  and the stack.
+**/
+typedef
+VOID
+(*DUMMY_VOID_FUNCTION_FOR_DATA_TEST)(
+  VOID
+  );
+
+/// ================================================================================================
+/// ================================================================================================
+///
+/// PRE REQ FUNCTIONS
+///
+/// ================================================================================================
+/// ================================================================================================
+
+/**
+  This function checks if the MM page guard policy is active for the target memory type within Context. Testing
+  page guards currently requires that buffers of the relevant memory type can be allocated. If the memory type
+  is Conventional or Persistent, the test will be skipped as we cannot allocate those types.

```

Comment:
Update to include the Unacceptable memory type?

Reply:
Updated